src/ameba/rule/typing/method_param_type_restriction.cr

```diff
@@ -0,0 +1,102 @@
+module Ameba::Rule::Typing
+  # A rule that enforces method parameters have type restrictions, with optional enforcement of block parameters
+  #
+  # For example, these are considered valid:
+  #
+  # ```
+  # def listen(a : String, b : String) : String
+  #   a + b
+  # end
+  #
+  # def hello(name : String?)
+  #   puts "Hello, " + (name || "there") + "!"
+  # end
+  # ```
+  #
+  # And these are considered invalid:
+  #
+  # ```
+  # def listen(a, b) : String
+  #   a + b
+  # end
+  #
+  # def hello(name)
+  #   puts "Hello, " + (name || "there") + "!"
+  # end
+  # ```
+  #
+  # When the config options `PrivateMethods` and `ProtectedMethods`
+  # are true, this rule is also applied to private and protected methods, respectively.
+  #
+  # The `BlockParam` configuration option will extend this to block params, where this is valid:
+  #
+  # ```
+  # def exec(&block : String -> String)
+  #   yield "cmd"
+  # end
+  # ```
+  #
+  # And these are invalid:
+  #
+  # ```
+  # def exec(&)
+  # end
+  #
+  # def exec(&block)
+  # end
+  # ```
+  #
+  # The config option `Undocumented` controls whether this rule applies to undocumented methods and methods with a `:nodoc:` directive.
+  #
+  # The config option `DefaultValue` controls whether this rule applies to parameters that have a default value.
+  #
+  # YAML configuration example:
+  #
+  # ```
+  # Typing/MethodParamTypeRestriction:
+  #   Enabled: false
+  #   DefaultValue: false
```

Comment:
**question:** Why is this disabled by default? A default value certainly gives a hint about what kind of value can be provided. But it provides much less information than a type restriction.